File: qa-portal/backend/app/models/project.py
from sqlalchemy import Column, String, Text, Boolean, Integer, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime, timezone
import uuid
import enum

from app.database import Base


class TestCaseStatus(str, enum.Enum):
    active = "active"
    inactive = "inactive"


class Project(Base):
    __tablename__ = "projects"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(150), nullable=False)
    description = Column(Text, nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    creator = relationship("User", foreign_keys=[created_by])
    modules = relationship("Module", back_populates="project", cascade="all, delete-orphan", order_by="Module.order")
    executions = relationship("TestExecution", back_populates="project", cascade="all, delete-orphan")


class Module(Base):
    __tablename__ = "modules"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(150), nullable=False)
    description = Column(Text, nullable=True)
    order = Column(Integer, default=0, nullable=False)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    project = relationship("Project", back_populates="modules")
    test_cases = relationship("TestCase", back_populates="module", cascade="all, delete-orphan")


class TestCase(Base):
    __tablename__ = "test_cases"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    module_id = Column(UUID(as_uuid=True), ForeignKey("modules.id", ondelete="CASCADE"), nullable=False)
    title = Column(String(255), nullable=False)
    preconditions = Column(Text, nullable=True)
    postconditions = Column(Text, nullable=True)
    status = Column(SAEnum(TestCaseStatus), nullable=False, default=TestCaseStatus.active)
    order = Column(Integer, default=0, nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    module = relationship("Module", back_populates="test_cases")
    creator = relationship("User", foreign_keys=[created_by])
    steps = relationship("TestStep", back_populates="test_case", cascade="all, delete-orphan", order_by="TestStep.order")

    @property
    def module_name(self) -> str:
        return self.module.name if self.module else ""


class TestStep(Base):
    __tablename__ = "test_steps"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    test_case_id = Column(UUID(as_uuid=True), ForeignKey("test_cases.id", ondelete="CASCADE"), nullable=False)
    order = Column(Integer, nullable=False)
    action = Column(Text, nullable=False)
    test_data = Column(Text, nullable=True)
    expected_result = Column(Text, nullable=False)

    test_case = relationship("TestCase", back_populates="steps")

File: qa-portal/frontend/lib/core/models/project.dart
class AppProject {
  final String id;
  final String name;
  final String description;
  final bool isActive;
  final String? createdAt;

  const AppProject({
    required this.id,
    required this.name,
    required this.description,
    required this.isActive,
    this.createdAt,
  });

  factory AppProject.fromJson(Map<String, dynamic> json) {
    return AppProject(
      id: json['id'] as String,
      name: json['name'] as String,
      description: (json['description'] as String?) ?? '',
      isActive: json['is_active'] as bool? ?? true,
      createdAt: json['created_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'is_active': isActive,
      'created_at': createdAt,
    };
  }
}

class AppModule {
  final String id;
  final String projectId;
  final String name;
  final String? description;
  final int order;
  final String? createdAt;

  const AppModule({
    required this.id,
    required this.projectId,
    required this.name,
    this.description,
    required this.order,
    this.createdAt,
  });

  factory AppModule.fromJson(Map<String, dynamic> json) {
    return AppModule(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      order: json['order'] as int,
      createdAt: json['created_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'project_id': projectId,
      'name': name,
      'description': description,
      'order': order,
      'created_at': createdAt,
    };
  }
}

class TestCase {
  final String id;
  final String moduleId;
  final String title;
  final String? preconditions;
  final String? postconditions;
  final String status;
  final List<TestStep> steps;
  final String? createdAt;

  const TestCase({
    required this.id,
    required this.moduleId,
    required this.title,
    this.preconditions,
    this.postconditions,
    required this.status,
    required this.steps,
    this.createdAt,
  });

  factory TestCase.fromJson(Map<String, dynamic> json) {
    return TestCase(
      id: json['id'] as String,
      moduleId: json['module_id'] as String,
      title: json['title'] as String,
      preconditions: json['preconditions'] as String?,
      postconditions: json['postconditions'] as String?,
      status: json['status'] as String,
      steps: (json['steps'] as List<dynamic>? ?? [])
          .map((s) => TestStep.fromJson(s as Map<String, dynamic>))
          .toList(),
      createdAt: json['created_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'module_id': moduleId,
      'title': title,
      'preconditions': preconditions,
      'postconditions': postconditions,
      'status': status,
      'steps': steps.map((s) => s.toJson()).toList(),
      'created_at': createdAt,
    };
  }
}

class TestStep {
  final String id;
  final String testCaseId;
  final int order;
  final String action;
  final String? testData;
  final String expectedResult;

  const TestStep({
    required this.id,
    required this.testCaseId,
    required this.order,
    required this.action,
    this.testData,
    required this.expectedResult,
  });

  factory TestStep.fromJson(Map<String, dynamic> json) {
    return TestStep(
      id: json['id'] as String,
      testCaseId: json['test_case_id'] as String,
      order: json['order'] as int,
      action: json['action'] as String,
      testData: json['test_data'] as String?,
      expectedResult: json['expected_result'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'test_case_id': testCaseId,
      'order': order,
      'action': action,
      'test_data': testData,
      'expected_result': expectedResult,
    };
  }
}
